pkg/virt-controller/watch/vmi.go

```diff
@@ -603,6 +604,9 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
 					}
 				}
 			}
+
+			vmiCopy.Status.EvictionStrategy = migrations.VMIEvictionStrategy(c.clusterConfig, vmi)
```

Comment:
Instead of introducing a new status field, would it be better to default the EvictionStrategy to `kubevirtCR.spec.evictionStrategy` if `vmi.spec.evictionStrategy` is empty during vmi creation?

vmiCreation happens once, this will make sure the correct strategy is be persisted at creation time. Here is an example of another field which does the same: https://github.com/kubevirt/kubevirt/blob/25d2e3329577d90a764d127b947e2608de28871f/pkg/virt-api/webhooks/defaults.go#L215, introduced in this PR: https://github.com/kubevirt/kubevirt/pull/8293

Reply:
I guess this will change the behavior: now if a `vmi.spec.evictionStrategy` is not set, the **current** value from KubeVirt CR is used. If the VMI value is defaulted, then it will be fixed as if the user set it on creation time.

Reply:
@vasiliy-ul the defaulting will only occur  if vmi.spec.evictionStrategy is not set. It's the same behavior. Thats what happens with `vmi.spec.architecture` as well. So I dont think it will be a behavior change

Reply:
@alaypatel07 yes I think this is a great idea! Thank you for the suggestion, I will update the PR.

Reply:
> So I dont think it will be a behavior change

Still, I am not sure. Looking at the code:

https://github.com/kubevirt/kubevirt/blob/0a91f46054c6429cd4311e4ecc255bdefe442b2d/pkg/util/migrations/migrations.go#L65-L71


So, the current flow:
- let's say, initially the eviction strategy in KubeVirt CR is `EvictionStrategyNone`
- I create a VM with an empty eviction strategy -> at this point in time it will use the strategy from the CR `EvictionStrategyNone`
- I change the eviction strategy in KubeVirt CR to `EvictionStrategyLiveMigrate`
- the VM will use the strategy from the CR, i.e. `EvictionStrategyLiveMigrate`

With the suggested defaulting, the behavior changes:
- initially the eviction strategy in KubeVirt CR is `EvictionStrategyNone`
- I create a VM with an empty eviction strategy -> `vmi.spec.evictionStrategy` is defaulted to `EvictionStrategyNone`
- I change the eviction strategy in KubeVirt CR to `EvictionStrategyLiveMigrate`
- the VM will still use `EvictionStrategyNone` since it is defaulted in the VMI spec.

Do I miss smth?

Reply:
> > So I dont think it will be a behavior change
> 
> Still, I am not sure. Looking at the code:
> 
> https://github.com/kubevirt/kubevirt/blob/0a91f46054c6429cd4311e4ecc255bdefe442b2d/pkg/util/migrations/migrations.go#L65-L71
> 
> So, the current flow:
> 
> * let's say, initially the eviction strategy in KubeVirt CR is `EvictionStrategyNone`
> * I create a VM with an empty eviction strategy -> at this point in time it will use the strategy from the CR `EvictionStrategyNone`
> * I change the eviction strategy in KubeVirt CR to `EvictionStrategyLiveMigrate`
> * the VM will use the strategy from the CR, i.e. `EvictionStrategyLiveMigrate`

It is not really like this because of one detail, let's say you specify `evictionStrategy` to `LiveMigrate` at cluster-level, you start the VM and then you later change the KV CR field to `None`, if you try to evict the VM the virt-launcher pod won't still be evicted as the corresponding PDB won't be deleted by virt-controller in reaction to the `evictionStrategy` change (as it should be), hence practically the current behaviour retains the initial `evictionStrategy`. To actually change it you have to restart the VM like most things in the VM object.

That said you actually uncovered a bug that would be fixed by using the `spec` instead of the `status` for storing the initial `evictionStrategy` as right now in virt-controller if we face the situation I just described the `evictionStrategy` for the VM will be evaluated to `None` after the CR change even if it actually behaves as `LiveMigrate`.

TLDR: using the spec is definitely the correct choice here both in terms of observability and behaviour.  

> 
> With the suggested defaulting, the behavior changes:
> 
> * initially the eviction strategy in KubeVirt CR is `EvictionStrategyNone`
> * I create a VM with an empty eviction strategy -> `vmi.spec.evictionStrategy` is defaulted to `EvictionStrategyNone`
> * I change the eviction strategy in KubeVirt CR to `EvictionStrategyLiveMigrate`
> * the VM will still use `EvictionStrategyNone` since it has been defaulted in the VMI spec.

This is indeed the behaviour we want.

Reply:
Hm, I see. Probably I was mostly looking at it from the API perspective without taking into account other aspects. Thank you for clarifying it!

Reply:
+1, this is the exact behavior we wanted with vmi.spec.architecture and its the right way to default optional fields that dont change over the lifetime of the vmi IMO